alert_words: Migrate Alert words to Watched phrases.

roanster007 opened the pull request:
This commit migrates alert words to watched phrases, through
the introduction to a new endpoint - users/me/watched_phrases.

It also renames the search filter from is:alerted to is:watched.

[api design thread](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Migrating.20alert.20words.20to.20watched.20phrases/near/1784780)

Fixes part of #28779

<!-- Describe your pull request here.-->



<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

andersk commented on zerver/lib/alert_words.py:
```diff
@@ -46,25 +56,30 @@ def get_alert_word_automaton(realm: Realm) -> ahocorasick.Automaton:
 
-def user_alert_words(user_profile: UserProfile) -> List[str]:
-    return list(AlertWord.objects.filter(user_profile=user_profile).values_list("word", flat=True))
+def user_alert_words(user_profile: UserProfile) -> List[Dict[str, str]]:
```
Should be `List[AlertWordData]`. Pydantic models can be serialized as well as deserialized—there should be no need to manually construct dicts for them.

andersk commented on zerver/actions/alert_words.py:
```diff
@@ -1,16 +1,16 @@
 
 
-def notify_alert_words(user_profile: UserProfile, words: Sequence[str]) -> None:
+def notify_alert_words(user_profile: UserProfile, words: Sequence[Dict[str, str]]) -> None:
     event = dict(type="alert_words", alert_words=words)
```
```suggestion
def notify_alert_words(user_profile: UserProfile, words: Sequence[AlertWordData]) -> None:
    event = dict(
        type="alert_words",
        alert_words=TypeAdapter(Sequence[AlertWordData]).dump_python(words),
    )
```

(Eventually as we migrate the event dicts themselves to Pydantic models, this `TypeAdapter` will become unnecessary here.)

andersk commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,15 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 244**
+
+* [`POST /register`](/api/register-queue), [`POST /events`](/api/get-events),
+  [`GET /users/me/alert_words`](/api/get-alert-words),
+  [`POST /users/me/alert_words`](/api/add-alert-words),
+  [`DELETE /users/me/alert_words`](/api/remove-alert-words): Migrate alert words
+  from list of strings to list of objects, with `word` as the property containing
+  the alert word string.
```
Are we comfortable making this change in a backwards-incompatible way?

roanster007 committed:
zerver: Migrate alert words to watched phrases.

This commit migrates the server from alert words to watched
phrases. A new endpoint - "users/me/watched_phrases" is
introduced as a part of this migration which uses pydantic
models.

In addition to that, alert_words is replaced with watched_
phrases, and alerted to watched in several files.

roanster007 committed:
web: Migrate alert words to watched phrases.

This commit migrates the web client to use the new api --
"users/me/watched_phrases", and new "watched_phrases"
event.

It also re names alert_words to watched_phrases, and
alerted to watched in several other files.

roanster007 committed:
search: Rename the search filter is:alerted to is:watched.

This commit renames the search filter "is:alerted" to
"is:watched".

This is a part of the migration of alert words to watched
phrases.

prakhar1144 commented on zerver/actions/alert_words.py:
```diff
@@ -1,20 +1,28 @@
-def notify_alert_words(user_profile: UserProfile, words: Sequence[str]) -> None:
-    event = dict(type="alert_words", alert_words=words)
+def notify_watched_phrases(user_profile: UserProfile, phrases: Sequence[WatchedPhraseData]) -> None:
+    event = dict(type="watched_phrases", watched_phrases=[w.dict() for w in phrases])
```
`for phrase in phrases`

prakhar1144 commented on zerver/lib/alert_words.py:
```diff
@@ -2,81 +2,121 @@
+class WatchedPhraseData(BaseModel):
+    model_config = ConfigDict(extra="forbid", frozen=True)
+
+    watched_phrase: Annotated[str, StringConstraints(max_length=100)]
```
I think it'd be better to name this simply `phrase`, so we can simply refer to it as `WatchedPhraseData.phrase`.

and in future.. `WatchedPhraseData.automatically_follow_topics`

prakhar1144 commented on zerver/actions/alert_words.py:
```diff
@@ -1,20 +1,28 @@
 
-def notify_alert_words(user_profile: UserProfile, words: Sequence[str]) -> None:
-    event = dict(type="alert_words", alert_words=words)
+def notify_watched_phrases(user_profile: UserProfile, phrases: Sequence[WatchedPhraseData]) -> None:
```
I'm thinking, should we use the term `phrase`/`phrases` only for the string, and not for `WatchedPhraseData` object. Currently, the terms aren't used consistently in the PR.

prakhar1144 commented on zerver/lib/alert_words.py:
```diff
@@ -2,81 +2,121 @@
+            "watched_phrase", flat=True
+        )
+    ]
+    return watched_phrase_data
```
`watched_phrases_data`

prakhar1144 commented on zerver/models/alert_words.py:
```diff
@@ -15,25 +15,25 @@ class AlertWord(models.Model):
     user_profile = models.ForeignKey(UserProfile, on_delete=CASCADE)
     # Case-insensitive name for the alert word.
-    word = models.TextField()
+    watched_phrase = models.TextField()
```
Should we call it just `phrase` ?

`phrase` field of `WatchedPhrase` model. It's clear enough, I think.

prakhar1144 commented:
Broadly looks great.

Not a complete review, but I think reviewing it would be much easier if we follow a pattern while renaming. i.e. Use the term `phrase` only for the string. `watched_phrase`/`watched_phrase_data` for the object. -- and not use them interchangebly.
